src/librbd/internal.cc

```diff
@@ -1635,6 +1636,216 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
+    CephContext *cct((CephContext *)io_ctx.cct());
+    ldout(cct, 20) << "trash_purge " << &io_ctx << " " << image_id << dendl;
+
+    int r = remove(io_ctx, "", image_id, prog_ctx, false, true);
+    if (r < 0) {
+      lderr(cct) << "error removing image " << image_id
+                 << ", which is pending deletion" << dendl;
+      return r;
+    }
+    r = cls_client::trash_remove(&io_ctx, image_id);
+    if (r < 0 && r != -ENOENT) {
+      lderr(cct) << "error removing image " << image_id
+                 << " from rbd_trash object" << dendl;
+      return r;
+    }
+    return 0;
+  }
+
+  int trash_restore(librados::IoCtx &io_ctx, const std::string &image_id,
+                    const std::string &image_new_name) {
+    CephContext *cct((CephContext *)io_ctx.cct());
+    ldout(cct, 20) << "trash_restore " << &io_ctx << " " << image_id << " "
+                   << image_new_name << dendl;
+
+    cls::rbd::TrashImageSpec trash_spec;
+    int r = cls_client::trash_get(&io_ctx, image_id, &trash_spec);
+    if (r < 0) {
+      lderr(cct) << "error getting image id " << image_id
+                 << " info from trash: " << cpp_strerror(r) << dendl;
+      return r;
+    }
+
+    std::string image_name = image_new_name;
+    if (image_name.empty()) {
+      // if user didn't specify a new name, let's try using the old name
+      image_name = trash_spec.name;
+      ldout(cct, 20) << "restoring image id " << image_id << " with name "
+                     << image_name << dendl;
+    }
+
+    // check if no image exists with the same name
+    bool create_id_obj = true;
+    std::string existing_id;
+    r = cls_client::get_id(&io_ctx, util::id_obj_name(image_name), &existing_id);
+    if (r < 0 && r != -ENOENT) {
+      lderr(cct) << "error checking if image " << image_name << " exists: "
+                 << cpp_strerror(r) << dendl;
+      return r;
+    } else if (r != -ENOENT){
+      // checking if we are recovering from an incomplete undelete
+      if (existing_id != image_id) {
+        ldout(cct, 2) << "an image with the same name already exists" << dendl;
+        return -EEXIST;
+      }
+      create_id_obj = false;
+    }
+
+    if (create_id_obj) {
+      ldout(cct, 2) << "adding id object" << dendl;
+      librados::ObjectWriteOperation op;
+      op.create(true);
+      cls_client::set_id(&op, image_id);
+      r = io_ctx.operate(util::id_obj_name(image_name), &op);
+      if (r < 0) {
+        lderr(cct) << "error adding id object for image " << image_name
+                   << ": " << cpp_strerror(r) << dendl;
+        return r;
+      }
+    }
+
+    ldout(cct, 2) << "adding rbd image from v2 directory..." << dendl;
+    r = cls_client::dir_add_image(&io_ctx, RBD_DIRECTORY, image_name,
+                                  image_id);
+    if (r < 0 && r != -EEXIST) {
+      lderr(cct) << "error adding image to v2 directory: "
+                 << cpp_strerror(r) << dendl;
+      return r;
+    }
```

Comment:
We don't need to handle the case that `r == -EEXIST` because if the rbd_directory already has an entry then it means that we are recovering from a failed restore